src/merkle.rs

```diff
@@ -41,33 +41,70 @@ pub const BUILD_DATA_BLOCK_SIZE: usize = 64 * BUILD_CHUNK_NODES;
 ///
 /// Merkle root is always the last element in the array.
 ///
-/// The number of inputs is not always a power of two which results in a
-/// balanced tree structure as above.  In that case, parent nodes with no
-/// children are also zero and parent nodes with only a single left node
-/// are calculated by concatenating the left node with itself before hashing.
-/// Since this function uses nodes that are pointers to the hashes, empty nodes
-/// will be nil.
+/// The number of inputs must always be a power of two.
+///
+/// This tree structure can consist of at most 3 layers of trees (of
+/// arity U, N and R, from bottom to top).
+///
+/// This structure ties together multiple Merkle Trees and allows
+/// supported properties of the Merkle Trees across it.  The
+/// significance of this class is that it allows an arbitrary number
+/// of sub-trees to be constructed and proven against.
+///
+/// To show an example, this structure can be used to create a single
+/// tree composed of 3 sub-trees, each that have branching factors /
+/// arity of 4.  Graphically, this may look like this:
+///
+/// ```text
+///                O
+///       ________/|\_________
+///      /         |          \
+///     O          O           O
+///  / / \ \    / / \ \     / / \ \
+/// O O  O  O  O O  O  O   O O  O  O
+///
+///
+/// At most, one more layer (top layer) can be constructed to group a
+/// number of the above sub-tree structures (not pictured).
 ///
-/// TODO: Ord
+/// BaseTreeArity is the arity of the base layer trees [bottom].
+/// SubTreeArity is the arity of the sub-tree layer of trees [middle].
+/// TopTreeArity is the arity of the top layer of trees [top].
+///
+/// With N and R defaulting to 0, the tree performs as a single base
+/// layer merkle tree without layers (i.e. a conventional merkle
+/// tree).
+#[allow(clippy::type_complexity)]
 #[derive(Debug, Clone, Eq, PartialEq, Default)]
-pub struct MerkleTree<T, A, K, U = U2>
+pub struct MerkleTree<E, A, S, BaseTreeArity = U2, SubTreeArity = U0, TopTreeArity = U0>
 where
-    T: Element,
-    A: Algorithm<T>,
-    K: Store<T>,
-    U: Unsigned,
+    E: Element,
+    A: Algorithm<E>,
+    S: Store<E>,
+    BaseTreeArity: Unsigned,
+    SubTreeArity: Unsigned,
+    TopTreeArity: Unsigned,
 {
-    data: K,
+    data: Option<S>,
     leafs: usize,
+    len: usize,
     height: usize,
 
+    top_layer_nodes: usize, // TopTreeArity
+    trees: Option<Vec<MerkleTree<E, A, S, BaseTreeArity, SubTreeArity>>>,
+
+    sub_tree_layer_nodes: usize, // SubTreeArity
+    sub_trees: Option<Vec<MerkleTree<E, A, S, BaseTreeArity>>>,
```

Comment:
Maybe instead of using three `Option`s this inner state could be simplified ab it nicer to make the available states more clear

```rust
enum Data {
	/// Single Tree
    Single(S),
    /// TODO: better name
	C1 {
	    trees: Vec<MerkleTree>>,
    }
	/// TODO: better name
	C2 {
	    trees: Vec<MerkleTree>>,
		sub_trees: Vec<MerkleTree>>,
    }
}

struct MerkleTree {
	data: Data,
	.. 
}
```